Make backups of `bagel.csv` and check it when tracking pipelines (+ refactor status -> doughnut)

michellewang opened the pull request:
This is not quite done yet but I'm creating the PR so that I don't forget about it.

Restore lost changes:
- Load old `bagel.csv` if it exists and replace rows as needed
  - [x] Currently this only works at the pipeline level, need to extend it so that it replaces only the specified session (since we now have a `session_id` argument when before we always tracked all sessions)
- Fix bug where all subjects were assumed to have all sessions, which is not the case

Closes #159.

nikhil153 commented on nipoppy/workflow/utils.py:
```diff
@@ -11,8 +11,10 @@
 # directory/file names
 DNAME_BACKUPS_MANIFEST = '.manifests'
 DNAME_BACKUPS_STATUS = '.doughnuts'
+DNAME_BACKUPS_BAGELS = '.bagels'
 FNAME_MANIFEST = 'manifest.csv'
 FNAME_STATUS = 'doughnut.csv'
+FNAME_BAGEL = 'bagel.csv'
```
maybe rename `STATUS` to `DOUGHNUT` to be consistent?

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -43,47 +48,76 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
-            _df["session"] = session
-            _df["pipeline_name"] = pipeline        
+        # Grab BIDS participants from the doughnut
+        doughnut_file = f"{DATASET_ROOT}/scratch/raw_dicom/doughnut.csv"
```
import doughnut.csv name from utils?

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -43,47 +48,76 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
+    for pipeline in pipelines:
+        pipe_tracker = Tracker(global_configs, dash_schema_file, pipeline) 
+        
+        mr_proc_root_dir, _, version = pipe_tracker.get_global_configs()
```
rename `mr_proc` :)

michellewang replied:
😅

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -43,47 +48,76 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
             _df["pipeline_version"] = version
+            _df["has_mri_data"] = TRUE # everyone in the status file has MRI data
+
+            tracker_csv = Path(mr_proc_root_dir, 'derivatives', FNAME_BAGEL)
```
rename `mr_proc` :)

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -43,47 +48,76 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
+                old_participants_session = set(old_proc_status_df_session[COL_BIDS_ID_MANIFEST])
+                old_pipelines_session = set(old_proc_status_df_session['pipeline_name'])
+                
+                # make sure the number of participants is consistent across pipelines
```
Not sure if this is before or after our discussion on querying doughnut regardless a full run or selective subset of participant run. But to me it should be possible to ensure number of participants is consistent across pipelines automatically. 
Th runtime error is likely to confuse the user more and trigger slow full runs unnecessarily.

michellewang replied:
This was from before the discussion yesterday. I didn't add the feature where the user can specify specific subjects to run yet because I think it's better to have it in another PR. I think we can address the consistent participant thing in that other PR also? The only thing that that doesn't sit too well with me is that we might have out-of-sync tracking info for different pipelines for the same participants (ex: fMRIPrep is up-to-date but MRIQC is not because for some reason it wasn't run). So that's why I had the error, though I could change it to a warning for now

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -97,29 +131,57 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
+            # don't write a new file if no changes
+            try:
+                if len(proc_status_df.compare(old_proc_status_df_full)) == 0:
+                    logger.info(f'\nNo change for pipeline {pipeline}, session {session}')
```
No change in status? Or `participant_id` x `session` combinations?

michellewang commented:
@nikhil153 addressed all your comments (I think). Also refactored some doughnut-related variable/function names which is why there are more files changes now

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -97,29 +133,57 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id=1,
                         _df.loc[bids_id,name] = status
                         _df.loc[bids_id,"pipeline_starttime"] = get_start_time(subject_dir)
-                        _df.loc[bids_id,"pipeline_endtime"] = UNAVAILABLE # TODO
+                        _df.loc[bids_id,"pipeline_endtime"] = get_end_time(subject_dir)
```
This could get very slow depending on number of files. Maybe we should either only check the files listed in the tracker config or parse the pipeline log to get completion timestamp.
